### DISCUSSION TITLE
Cluster forward/deferred rendering?

### DISCUSSION THREAD

**Discussion by swq0553** [rokups/rbfx#422] [General]
Has anyone tried to implement a “Cluster forward/deferred rendering” renderpath？

**Comment by eugeneko** [comment]
@JSandusky did you maybe tried this in your journey?
I actually wanted to try implementing Forward+ at some point but it's quite low priority for me right now.

**Comment by JSandusky** [comment]
I've got a bunch of old code that does the cluster assignment on the CPU but I've never done it fully on the GPU. I eventually went with BVH Accelerated Shading, doing a Drobot style light mask in a uint4 as part of a programmable pulling pipe because that's more suitable for VR (worldspace instead of view-space per eye).

Here's that stuff  (CPU clustering just crude box tests [it's not Urho code, math types are MathGeoLib so should be trivial] and the BVH sphere tree I use) https://gist.github.com/JSandusky/25750e06fa945a7df75a199c4ad4ad26

Urho/RBFX-wise without adding shadowmap atlasing it could only really be done by identifying trivial lights that can be represented analytically (which is probably enough for most real cases) with non-analytic lights shading same as usual. The distinction I use to identify that is:

```cpp
bool Light::IsAnalytic() const
{
    if (auto renderer = GetSubsystem<Renderer>())
    {
        // must use defaults, which we can reproduce with math in the shader
        if (renderer->GetDefaultLightRamp() != rampTexture_ ||
            renderer->GetDefaultLightSpot() != shapeTexture_)
            return false;
    }

    // Although analytic it doesn't make sense for a directional light to not cast-shadows except
    // in some really odd situations (non-photoreal)
    if (lightType_ == LIGHT_DIRECTIONAL)
        return false;

    // Must not cast shadows to be analytic
    return !GetCastShadows();
}
```

I'm too bogged down in XR to give it a go in RBFX anytime soon.
